### PR TITLE
feat(Core/Items): Delete old item id from recovery item

### PR DESCRIPTION
**Pull request by IntelligentQuantum**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been: I didn't test this code and i dont know if it works or not xD
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [x] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. enable this option from world conf ItemDelete.Method = 1
2. sell your items to the vendor
3. and enable this option ItemDelete.KeepDays = 30, items should deleted after 30day from recovery_item

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/apps/worldserver/worldserver.conf.dist [hunk 1]
```diff
@@ -2602,6 +2602,15 @@ ItemDelete.Quality = 3
...
 ItemDelete.ItemLevel = 80
 
+#
+#
+#    ItemDelete.KeepDays
+#        Description: Time (in days)
+#        Default:     30 - (Enabled)
+#                     0  - (Disabled, Don't delete any items)
+
+ItemDelete.KeepDays = 30
```
Comment (elthehablo) [inline src/server/apps/worldserver/worldserver.conf.dist:2612]: Default 0 to not leave anyone who didn't read this PR surprised when items are permanently deleted?

**Review by elthehablo — APPROVED** [review]

File: data/sql/updates/pending_db_characters/rev_1707981003566465200.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+--
+ALTER TABLE `recovery_item`
+    ADD COLUMN `DeleteDate` INT(10) UNSIGNED NULL DEFAULT NULL AFTER `Count`;
```
Comment (Helias) [inline data/sql/updates/pending_db_characters/rev_1707981003566465200.sql:3]: it's good to add a newline at the end of the file, moreover `INT(10)` is not necessary anymore, you can put just `INT` ( [source](https://stackoverflow.com/questions/58938358/mysql-warning-1681-integer-display-width-is-deprecated) ).

**Review by Helias — APPROVED** [review]

File: src/server/game/Entities/Player/Player.cpp [hunk 3]
```diff
@@ -4323,6 +4323,49 @@ void Player::DeleteOldCharacters(uint32 keepDays)
     }
 }
 
+/**
+ * Items which were kept back in the database after being deleted and are now too old (see config option "ItemDelete.KeepDays"), will be completely deleted.
+ */
+void Player::DeleteOldItems()
```
Comment (walkline) [inline src/server/game/Entities/Player/Player.cpp:4329]: What do you think about renaming this and others to something like `DeleteOldRecoveryItems`? Without context this function definition may be misleading.

File: src/server/game/World/World.cpp [hunk 4]
```diff
@@ -1265,6 +1265,7 @@ void World::LoadConfigSettings(bool reload)
     _bool_configs[CONFIG_ITEMDELETE_VENDOR]    = sConfigMgr->GetOption<bool>("ItemDelete.Vendor", 0);
     _int_configs[CONFIG_ITEMDELETE_QUALITY]    = sConfigMgr->GetOption<int32>("ItemDelete.Quality", 3);
     _int_configs[CONFIG_ITEMDELETE_ITEM_LEVEL] = sConfigMgr->GetOption<int32>("ItemDelete.ItemLevel", 80);
+    _int_configs[CONFIG_ITEMDELETE_KEEP_DAYS]  = sConfigMgr->GetOption<int32>("ItemDelete.KeepDays", 30);
```
Comment (walkline) [inline src/server/game/World/World.cpp:1268]: Maybe use 0 as default value?

**Comment by Nyeriah** [comment]
@IntelligentQuantum please address the reviews so we can approve merge the PR

File: src/server/database/Database/Implementation/CharacterDatabase.cpp [hunk 5]
```diff
@@ -585,11 +585,13 @@ void CharacterDatabaseConnection::DoPrepareStatements()
...
 
     // Recovery Item
-    PrepareStatement(CHAR_INS_RECOVERY_ITEM, "INSERT INTO recovery_item (Guid, ItemEntry, Count) VALUES (?, ?, ?)", CONNECTION_SYNCH);
+    PrepareStatement(CHAR_INS_RECOVERY_ITEM, "INSERT INTO recovery_item (Guid, ItemEntry, Count, DeleteDate) VALUES (?, ?, ?, UNIX_TIMESTAMP())", CONNECTION_SYNCH);
     PrepareStatement(CHAR_SEL_RECOVERY_ITEM, "SELECT id, itemEntry, Count, Guid FROM recovery_item WHERE id = ?", CONNECTION_SYNCH);
     PrepareStatement(CHAR_SEL_RECOVERY_ITEM_LIST, "SELECT id, itemEntry, Count FROM recovery_item WHERE Guid = ? ORDER BY id DESC", CONNECTION_SYNCH);
     PrepareStatement(CHAR_DEL_RECOVERY_ITEM, "DELETE FROM recovery_item WHERE Guid = ? AND ItemEntry = ? AND Count = ? ORDER BY Id DESC LIMIT 1", CONNECTION_ASYNC);
     PrepareStatement(CHAR_DEL_RECOVERY_ITEM_BY_RECOVERY_ID, "DELETE FROM recovery_item WHERE id = ?", CONNECTION_ASYNC);
+    PrepareStatement(CHAR_SEL_RECOVERY_ITEM_OLD_ITEMS, "SELECT Guid, ItemEntry FROM recovery_item WHERE DeleteDate IS NOT NULL AND DeleteDate < ?", CONNECTION_SYNCH);
+    PrepareStatement(CHAR_DEL_RECOVERY_ITEM_BY_GUID, "DELETE FROM recovery_item WHERE Guid = ?", CONNECTION_ASYNC);
```
[the remark this answers is not in the document]
Comment (IntelligentQuantum) [reply src/server/database/Database/Implementation/CharacterDatabase.cpp:594]: Yes, i wanna use it in others sections and websites (For Custom things)

File: src/server/apps/worldserver/worldserver.conf.dist [hunk 6]
```diff
@@ -2602,6 +2602,15 @@ ItemDelete.Quality = 3
...
 ItemDelete.ItemLevel = 80
 
+#
+#
+#    ItemDelete.KeepDays
+#        Description: Time (in days)
+#        Default:     30 - (Enabled)
+#                     0  - (Disabled, Don't delete any items)
+
+ItemDelete.KeepDays = 0
```
Comment (Kitzunu) [inline src/server/apps/worldserver/worldserver.conf.dist:2612]: You say the default is 30, but you set it to 0 🤔

File: src/server/apps/worldserver/worldserver.conf.dist [hunk 7]
```diff
@@ -2602,6 +2602,15 @@ ItemDelete.Quality = 3
 
 ItemDelete.ItemLevel = 80
 
+#
```
Comment (Kitzunu) [inline src/server/apps/worldserver/worldserver.conf.dist:2605]:
```suggestion
```

File: src/server/apps/worldserver/worldserver.conf.dist (hunk 6 above)
Comment (IntelligentQuantum) [reply src/server/apps/worldserver/worldserver.conf.dist]: 😨

**Review by sudlud — APPROVED** [review]

**Pull request merged**